packages/google_sign_in/lib/google_sign_in.dart

```diff
@@ -126,31 +119,80 @@ class GoogleSignIn {
+        : null;
     _streamController.add(_currentUser);
     return _currentUser;
   }
 
-  /// The currently signed in account, or null if the user is signed out
-  GoogleSignInAccount _currentUser;
+  /// Keeps track of the most recently scheduled method call.
+  _MethodCompleter _lastMethodCompleter;
+
+  /// Adds call to [method] in a queue for execution.
+  ///
+  /// At most one in flight call is allowed to prevent concurrent (out of order)
+  /// updates to [currentUser] and [onCurrentUserChanged].
+  Future<GoogleSignInAccount> _addMethodCall(String method) {
+    if (_lastMethodCompleter == null) {
+      _lastMethodCompleter = new _MethodCompleter(method);
+      _lastMethodCompleter.complete(_callMethod(method));
+      return _lastMethodCompleter.future;
+    }
+
+    // Return the same Future for consecutive calls to the same method.
+    if (_lastMethodCompleter.method == method &&
+        !_lastMethodCompleter.isCompleted) {
+      return _lastMethodCompleter.future;
+    }
+
+    var completer = new _MethodCompleter(method);
+    _lastMethodCompleter.future.whenComplete(() {
+      // If after the last completed call currentUser is not null and requested
+      // method is a sign in method, re-use the same authenticated user
+      // instead of making extra call to the native side.
+      const signInMethods = const ['signIn', 'signInSilently'];
+      if (signInMethods.contains(method) && _currentUser != null) {
+        completer.complete(_currentUser);
+      } else {
+        completer.complete(_callMethod(method));
+      }
+    }).catchError((_) {
+      // Ignore if previous call completed with an error.
+    });
+    _lastMethodCompleter = completer;
+    return _lastMethodCompleter.future;
+  }
+
+  /// The currently signed in account, or null if the user is signed out.
   GoogleSignInAccount get currentUser => _currentUser;
+  GoogleSignInAccount _currentUser;
 
   /// Attempts to sign in a previously authenticated user without interaction.
   ///
-  /// If there is already a signed-in user, returns it.
-  Future<GoogleSignInAccount> signInSilently() async => _currentUser ?? _callMethod('signInSilently');
+  /// Authentication flow is triggered only by the first call to this method,
+  /// any consequent calls resolve to the same user instance or `null`. When
+  /// `null` is returned this means no previously authenticated user exists
+  /// and interactive [signIn] is required.
+  ///
+  /// Authentication can be re-initiated only after [signOut] or [disconnect].
+  Future<GoogleSignInAccount> signInSilently() =>
+      _addMethodCall('signInSilently');
 
   /// Starts the sign-in process.
   ///
-  /// If there is already a signed-in user, returns it.
-  Future<GoogleSignInAccount> signIn() async => _currentUser ?? _callMethod('signIn');
+  /// Interactive sign-in flow is triggered only be the first call to this method,
+  /// any consequent calls resolve to the same user instance or `null`. When
```

Comment:
I realized I can't use the same pattern for "sign out" methods because `signOut` and `disconnect` perform different operations.

If a user was signed out by a call to `signOut` and we get a `disconnect` call we still need to call native side to revoke connection.